gfx.cpp

```diff
@@ -4540,6 +4561,41 @@ class GfxInternal
         return kGfxResult_NoError;
     }
 
+    void decayResourceState()
+    {
+        for(uint32_t i = 0; i < buffers_.size(); ++i)
+        {
+            auto *buffer = buffers_.at(i);
+            if(buffer != nullptr && buffer->initial_resource_state_ == D3D12_RESOURCE_STATE_COMMON && !buffer->transitioned_ && (buffer->resource_state_ & D3D12_RESOURCE_STATE_GENERIC_READ) == buffer->resource_state_)
+                buffer->resource_state_ = D3D12_RESOURCE_STATE_COMMON;
+        }
+        for(uint32_t i = 0; i < textures_.size(); ++i)
+        {
+            auto *texture = textures_.at(i);
+            if(texture != nullptr && !texture->transitioned_ && (texture->resource_state_ & D3D12_RESOURCE_STATE_ALL_SHADER_RESOURCE) == texture->resource_state_)
+                texture->resource_state_ = D3D12_RESOURCE_STATE_COMMON;
+        }
+        if(dbgCommandList != nullptr)
+        {
+            for(uint32_t i = 0; i < buffers_.size(); ++i)
+            {
+                auto *buffer = buffers_.at(i);
+                if(buffer != nullptr)
+                {
+                    dbgCommandList->AssertResourceState(buffer->resource_, D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES, buffer->resource_state_);
+                }
+            }
+            for(uint32_t i = 0; i < textures_.size(); ++i)
+            {
+                auto *texture = textures_.at(i);
+                if(texture != nullptr)
+                {
+                    dbgCommandList->AssertResourceState(texture->resource_, D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES, texture->resource_state_);
+                }
+            }
+        }
+    }
```

Comment:
Implicitly transitioned resources will also implicitly decay after a executeList call. This function will now auto change the resource states accordingly.
AssertResourceState is used here to check that the resources match what the internal state assumes. Its in its own loop to prevent performance hits when not debugging. This function will cause debug errors to be logged next time the command list is executed